corehq/project_limits/rate_counter/rate_counter.py

```diff
@@ -68,6 +68,13 @@ def get(self, scope, timestamp=None):
         contribution_from_earliest = earliest_grain_count * (1 - progress_in_current_grain)
         return sum(counts) + contribution_from_earliest
```

Comment:
I think a cleaner solution (continuing the comment I left on `retry_after`) would be to have a version of this same method called `get_count_and_wait_time(scope, timestamp=None, threshold=threshold)`, since the two values are calculated from the same underlying data, and it would be weird to pull it twice moments apart and get slightly different values.

`wait_time` would be the smallest (non-negative) value in seconds such that `self.get(scope, timestamp=timestamp + wait_time)` equals `threshold`. This can be calculated first in terms of the number of grains (decimal, not a whole number) and then that can be multiplied by `self.grain_duration` to get it in terms of seconds.

It would look something like this

```python
    def get_count_and_wait_time(self, scope, threshold, timestamp=None):
        if timestamp is None:
            timestamp = time.time()
        counts = [
            self.grain_counter.get(scope, timestamp - i * self.grain_duration,
                                   key_is_active=(i == 0))
            for i in range(self.grains_per_window + 1)
        ]
        earliest_grain_count = counts.pop()
        # This is the percentage of the way through the current grain we are
        progress_in_current_grain = (timestamp % self.grain_duration) / self.grain_duration
        # This is the count from the percentage of the earliest grain that should count
        contribution_from_earliest = earliest_grain_count * (1 - progress_in_current_grain)
        current_count = sum(counts) + contribution_from_earliest
        if current_count > threshold:
            # compute how many grains back we need to go for the count to equal the threshold
            cumulative_grains = 0
            cumulative_count = 0
            for count in (counts + earliest_grain_count):
                new_count = cumulative_count + count
                if threshold < new_count:
                    # cumulative_count <= threshold < cumulative_count + count
                    # solve for x: cumulative_count + x*count = threshold
                    # x is the proportion of the new grain we need to get through to exactly meet the threshold
                    cumulative_grains += (threshold - cumulative_count) / count
                    break
                else:
                    cumulative_count = new_count
                    cumulative_grains += 1
            wait_time = (self.grains_per_window - cumulative_grains) * self.grain_duration
        else:
            wait_time = 0
        return current_count, wait_time
````

writing that out, it's kind of a long function, so might be worth seeing if you can break it up into two pieces. The only thing they really need to share is the `counts = [...]` part where it fetches the counts for the grains for the current window. Since we may not always _want_/need the wait time, we should also probably retain both `.get` and this new method, and have them call the same private method to do the current count calculation.

I realize this all looks complicated and it's a big change. I'm hopeful that this is fairly straightforward to add to existing tests (like here https://github.com/dimagi/commcare-hq/blob/cc93bc6ab718bde08e8d16e374d5cc25eff94958/corehq/project_limits/tests/test_rate_counter.py#L66) though.

I won't insist on this, but it does seem more accurate/consistent to me, and with a fairly accurate estimate like this, I can imagine it being useful in other places as well.

Reply:
This is gold, thanks @dannyroberts. I'm going to merge this PR for the sake of unblocking other work, but I like your suggestions, so how about this:

I have a [small refactor PR](https://github.com/dimagi/commcare-hq/pull/33648) to try and make some things easier to follow in this area (at least for me). If we're adding your suggested change it will in anycase change that PRs approach, so let's introduce this change as part of that PR. I'll reach out to you for a pairing session.